grype/vex/openvex/implementation.go

```diff
@@ -0,0 +1,237 @@
+// identifiers identifying the scanned image.
+func productIDentifiersFromContext(pkgContext *pkg.Context) ([]string, error) {
+	switch v := pkgContext.Source.Metadata.(type) {
+	case source.StereoscopeImageSourceMetadata:
+		// TODO(puerco): We can create a wider definition here. This effectivley
+		// adds the multiarch image and the image of the OS running grype. We
+		// could generate more identifiers to match better.
+		return identifiersFromDigests(v.RepoDigests), nil
+	default:
+		// Fail for now
+		return nil, errors.New("source type not supported for VEX")
+	}
+}
+
+func identifiersFromDigests(digests []string) []string {
+	identifiers := []string{}
+
+	for _, d := range digests {
+		// The first identifier is the original image reference:
+		identifiers = append(identifiers, d)
+
+		// Now, parse the digest
+		parts := strings.SplitN(d, "@", 2)
+		if len(parts) != 2 {
+			continue
+		}
+
+		name := ""
+		repoURL := ""
+		digestString := strings.TrimPrefix(parts[1], "sha256:")
+		subparts := strings.Split(parts[0], "/")
+		switch len(subparts) {
+		case 1:
+			name = subparts[0]
+			repoURL = ""
+		default:
+			name = subparts[(len(subparts) - 1)]
+			repoURL = strings.Join(subparts[0:len(subparts)-1], "/")
+		}
+
+		if name == "" {
+			continue
+		}
+		qMap := map[string]string{}
+		// Add
+		if repoURL != "" {
+			qMap["repository_url"] = repoURL
+		}
+		qs := packageurl.QualifiersFromMap(qMap)
+		identifiers = append(identifiers, packageurl.NewPackageURL(
+			"oci", "", name, fmt.Sprintf("sha256%%3A%s", digestString), qs, "",
+		).String())
+
+		// TODO(puerco): Should also pass the digests only? They could be listed
+		// in the openvex document an foks may choose to vex on the hash.
+	}
+	return identifiers
+}
+
+// subcomponentIdentifiersFromMatch returns the list of identifiers from the
+// package where grype did the match.
+func subcomponentIdentifiersFromMatch(m *match.Match) []string {
+	ret := []string{}
+	if m.Package.PURL != "" {
+		ret = append(ret, m.Package.PURL)
+	}
+
+	// TODO(puerco):Implement CPE matching in openvex/go-vex
+	/*
+		for _, c := range m.Package.CPEs {
+			ret = append(ret, c.String())
+		}
+	*/
+	return ret
+}
+
+// FilterMatches takes a set of scanning results and moves any results marked in
+// the VEX data as fixed or not_affected to the ignored list.
+func (ovm *Processor) FilterMatches(
+	docRaw interface{}, ignoreRules []match.IgnoreRule, pkgContext *pkg.Context, matches *match.Matches, ignoredMatches []match.IgnoredMatch,
+) (*match.Matches, []match.IgnoredMatch, error) {
+	doc, ok := docRaw.(*openvex.VEX)
+	if !ok {
+		return nil, nil, errors.New("unable to cast vex document as openvex")
+	}
+
+	remainingMatches := match.NewMatches()
+
+	outfile, err := os.Create("/tmp/output.txt")
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to open output file: %w", err)
+	}
+	defer outfile.Close()
+
+	fmt.Fprintln(outfile, "VEX Debug Data")
```

Comment:
I can help out with UI changes to plug into the bus and update the existing components. I think the component will need to get refactored some to account for multiple events.